lib/motsu/src/context.rs

```diff
@@ -1,34 +1,97 @@
+/// Context of stylus unit tests associated with the current test thread.
+#[allow(clippy::module_name_repetitions)]
+pub struct TestContext {
+    thread_name: TestThreadName,
+}
+
+impl TestContext {
+    /// Get test context associated with the current test thread.
+    #[must_use]
+    pub fn current() -> Self {
+        Self { thread_name: TestThreadName::current() }
+    }
+
+    /// Get the value at `key` in storage.
+    pub(crate) fn get_bytes(self, key: &Bytes32) -> Bytes32 {
+        let storage = TESTS_STORAGE.entry(self.thread_name).or_default();
+        storage.contract.get(key).copied().unwrap_or_default()
+    }
+
+    /// Get the raw value at raw `key` in storage.
+    pub(crate) unsafe fn get_bytes_raw(self, key: *const u8, value: *mut u8) {
+        let key = read_bytes32(key);
+
+        write_bytes32(value, self.get_bytes(&key));
+    }
+
+    /// Set the value at `key` in storage to `val`.
+    pub(crate) fn set_bytes(self, key: Bytes32, val: Bytes32) {
+        let mut storage = TESTS_STORAGE.entry(self.thread_name).or_default();
+        storage.contract.insert(key, val);
+    }
+
+    /// Set the raw value at `key` in storage to raw `val`.
+    pub(crate) unsafe fn set_bytes_raw(self, key: *const u8, value: *const u8) {
+        let (key, value) = (read_bytes32(key), read_bytes32(value));
+        self.set_bytes(key, value);
+    }
+
+    /// Clears storage, removing all key-value pairs associated with the current
+    /// test thread.
+    pub fn reset_storage(self) {
+        TESTS_STORAGE.remove(&self.thread_name);
+    }
+}
+
+/// Storage mock: A global mutable key-value store.
+/// Allows concurrent access.
 ///
-/// See [`with_context`].
-pub(crate) static STORAGE_MUTEX: Mutex<()> = Mutex::new(());
-
-/// Acquires access to storage.
-pub fn acquire_storage() -> MutexGuard<'static, ()> {
-    STORAGE_MUTEX.lock().unwrap_or_else(|e| {
-        reset_storage();
-        e.into_inner()
-    })
+/// The key is the name of the test thread, and the value is the storage of the
+/// test case.
+static TESTS_STORAGE: Lazy<DashMap<TestThreadName, TestStorage>> =
+    Lazy::new(DashMap::new);
+
+/// Test thread name metadata.
+#[derive(Clone, Eq, PartialEq, Hash)]
+struct TestThreadName(String);
+
+impl TestThreadName {
+    /// Get the name of the current test thread.
+    fn current() -> Self {
+        let current_thread_name = std::thread::current()
+            .name()
+            .expect("should retrieve current thread name")
+            .to_string();
+        Self(current_thread_name)
+    }
 }
 
-/// Decorates a closure by running it with exclusive access to storage.
-#[allow(clippy::module_name_repetitions)]
-pub fn with_context<C: StorageType>(closure: impl FnOnce(&mut C)) {
-    let _lock = acquire_storage();
-    let mut contract = C::default();
-    closure(&mut contract);
-    reset_storage();
+/// Context of the test case.
+#[derive(Default)]
+struct TestStorage {
```

Comment:
In the nearest future it is possible to add to `TestStorage` struct fields like `msg_sender`, `msg_value` and so on for every mock data.